diff --git a/src/nfa/noncontiguous.rs b/src/nfa/noncontiguous.rs
index af32617..c732be8 100644
--- a/src/nfa/noncontiguous.rs
+++ b/src/nfa/noncontiguous.rs
@@ -1079,21 +1079,6 @@ impl<'a> Compiler<'a> {
                 "expected number of patterns to match pattern ID"
             );
             self.nfa.pattern_lens.push(patlen);
-            // We add the pattern to the prefilter here because the pattern
-            // ID in the prefilter is determined with respect to the patterns
-            // added to the prefilter. That is, it isn't the ID we have here,
-            // but the one determined by its own accounting of patterns.
-            // To ensure they line up, we add every pattern we see to the
-            // prefilter, even if some patterns ultimately are impossible to
-            // match (in leftmost-first semantics specifically).
-            //
-            // Another way of doing this would be to expose an API in the
-            // prefilter to permit setting your own pattern IDs. Or to just use
-            // our own map and go between them. But this case is sufficiently
-            // rare that we don't bother and just make sure they're in sync.
-            if self.builder.prefilter {
-                self.prefilter.add(pat);
-            }
 
             let mut prev = self.nfa.special.start_unanchored_id;
             let mut saw_match = false;
@@ -1145,6 +1130,10 @@ impl<'a> Compiler<'a> {
             // Once the pattern has been added, log the match in the final
             // state that it reached.
             self.nfa.add_match(prev, pid)?;
+            // BUG: add to the prefilter only *after* the pattern walk.
+            if self.builder.prefilter {
+                self.prefilter.add(pat);
+            }
         }
         Ok(())
     }
